### ISSUE TITLE
Unsigned Integer Overflow

### ISSUE THREAD

**Issue by snehashis-roy** [AdaptiveParticles/LibAPR#178]
While converting an uint16 image to APR format and converting back to tiff, it seems the some zero background pixels are arbitrarily changed to 65535. My images are uint16, with some pixels exactly 0.

Here are the commands I am using,
```
LibAPR/build/examples/Example_get_apr -i input.tif -o test -lambda 0
```
and
```
LibAPR/build/examples/Example_reconstruct_image -i test.apr -o test
```
When pixels are non-zero, the difference between my input.tif and output test_pc.tif are small, except for zero pixels being randomly changed to 65535.
I am happy to provide examples if this is not reproducible.

**Comment by joeljonsson** [comment]
Thanks for reporting this! 
What are the minimum and maximum values in your input image?
It sounds like there are underflows in the sampled values, but I don't see how that can happen in the actual sampling method. I will have to look into it in the morning. 

In the meantime, could you perhaps try changing [APRConverter<uint16_t>](https://github.com/AdaptiveParticles/LibAPR/blob/dca7e7fb053613ca327ced55f5cda245e5bd0d6a/examples/Example_get_apr.cpp#L51) to `APRConverter<float>`  in the Example_get_apr script, and see if that makes a difference?

**Comment by snehashis-roy** [comment]
The max is 65535, min 0. Here is a reproducible dataset if you prefer.

Input ```https://hpc.nih.gov/~roys5/1.tif```
Reconstructed ```https://hpc.nih.gov/~roys5/2_pc.tif```

**Comment by joeljonsson** [comment]
Alright, there appears to be several issues here:

1. The over/underflow in the reconstructed images are not due to random noise being added in `Example_reconstruct_image`. To disable this, set [this value](https://github.com/AdaptiveParticles/LibAPR/blob/dca7e7fb053613ca327ced55f5cda245e5bd0d6a/examples/Example_reconstruct_image.cpp#L157) to false. I'm guessing this is a leftover from some experimentation that somehow made it into the master branch. I will turn this in to an option which defaults to false.

2. Converting uint16 images with values across the entire range using `APRConverter<uint16_t>` should not be allowed, and has been addressed previously in #152. In your case this can cause errors in two ways: first, an offset is added to the input image which will overflow values near 65535. Second, the b-spline fit might overflow for values near the extremes after the offset has been added. An exception should be thrown here, but it appears that this doesn't happen when called through the example executable. When I try to convert your images from [pyapr](https://github.com/AdaptiveParticles/pyapr) it correctly results in an error. I will have to look over the error handling for this.

Either way, your best option will be to use `APRConverter<float>` for conversion, and disabling the random gitter in the reconstruction.

As a side note, it is quite tricky to get a good adaptation for your image. Using the `-auto_parameters` option seems to result in adaptation to the small bright region, and severe undersampling of the larger but dimmer region. If you want to play around with manual tuning of conversion parameters more efficiently, I would recommend installing [pyapr](https://github.com/AdaptiveParticles/pyapr) and running the [interactive conversion demo](https://github.com/AdaptiveParticles/pyapr/blob/master/demo/get_apr_interactive_demo.py).
